snow/networking/blacklist/blacklist.go

```diff
@@ -0,0 +1,231 @@
+	// RegisterResponse registers the response to a query message
+	RegisterResponse(ids.ShortID, uint32)
+	// QueryFailed registers that a query did not receive a response within our synchrony bound
+	QueryFailed(ids.ShortID, uint32)
+}
+
+// RegisterQuery attempts to register a query from [validatorID] and returns true
+// if that request should be made (not subject to blacklisting)
+func (b *queryBlacklist) RegisterQuery(validatorID ids.ShortID, requestID uint32) bool {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	key := validatorID.Key()
+	if blacklisted := b.blacklisted(validatorID); blacklisted {
+		return false
+	}
+	validatorRequests, ok := b.pendingQueries[key]
+	if !ok {
+		validatorRequests = make(map[uint32]struct{})
+		b.pendingQueries[key] = validatorRequests
+	}
+	validatorRequests[requestID] = struct{}{}
+
+	return true
+}
+
+// RegisterResponse removes the query from pending
+func (b *queryBlacklist) RegisterResponse(validatorID ids.ShortID, requestID uint32) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	if ok := b.removeQuery(validatorID, requestID); !ok {
+		return
+	}
+
+	// Reset consecutive failures on success
+	delete(b.consecutiveFailures, validatorID.Key())
+}
+
+// QueryFailed notes a failure and blacklists [validatorID] if necessary
+func (b *queryBlacklist) QueryFailed(validatorID ids.ShortID, requestID uint32) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	if ok := b.removeQuery(validatorID, requestID); !ok {
+		return
+	}
+
+	key := validatorID.Key()
+	// Add a failure and blacklist [validatorID] if it has
+	// passed the threshold
+	b.consecutiveFailures[key]++
+	if b.consecutiveFailures[key] >= b.threshold {
+		b.blacklist(validatorID)
+	}
+}
+
+func (b *queryBlacklist) blacklist(validatorID ids.ShortID) {
+	if b.blacklistSet.Contains(validatorID) {
+		return
+	}
+
+	key := validatorID.Key()
+
+	// Add to blacklist times
+	b.blacklistTimes[key] = b.clock.Time().Add(b.duration)
+	b.blacklistOrder.PushBack(validatorID)
+	b.blacklistSet.Add(validatorID)
+
+	// Note: there could be a memory leak if a large number of
+	// validators were added, sampled, and then were blacklisted
+	// and never sampled again. Due to the minimum staking amount
+	// and durations this is not a realistic concern.
+	b.cleanup()
+}
+
+// blacklisted checks if [validatorID] is currently blacklisted
+// and calls cleanup if its blacklist period has elapsed
+func (b *queryBlacklist) blacklisted(validatorID ids.ShortID) bool {
+	key := validatorID.Key()
+
+	end, ok := b.blacklistTimes[key]
+	if !ok {
+		return false
+	}
+
+	if b.clock.Time().Before(end) {
+		return true
+	}
+
+	// If a blacklisted item has expired, cleanup the blacklist
+	b.cleanup()
+	return false
+}
+
+// cleanup ensures that we have not blacklisted too much stake
+// and removes anything from the blacklist whose time has expired
+func (b *queryBlacklist) cleanup() {
+	currentWeight, err := b.vdrs.SubsetWeight(b.blacklistSet)
+	if err != nil {
+		b.reset()
```

Comment:
I don't think this should be a silent error...it should never happen, right?

Reply:
Yes, I'll add logging in here. It's possible for this to happen due to validator weight changes.